operations_on_datastructures/inorder_successor_of_bst.cpp

```diff
@@ -0,0 +1,407 @@
+ *
+ * #### Method 1: Use parent pointer (store the address of parent nodes)
+ *      If a node does not have right subtree, and we already visited the node
+ * itself, then the next node will be its parent node according to inorder
+ * traversal, and if we are going to parent from left, then the parent would be
+ * unvisited. In other words, go to the nearest ancestor for which given node
+ * would be in left subtree.
+ *
+ * #### Method 2: Search from the root node
+ *      In case if there is no link to the parent, we need to walk the tree
+ * starting from the root node to the given node, by doing so, we are visiting
+ * every ancestor of the given node. In order successor would be the deepest
+ * node in this path for which given node is in left subtree.
+ *
+ * @author [Nitin Sharma](https://github.com/foo290)
+ * */
+
+#include <cassert>   ///  For assert
+#include <iostream>  ///  For IO Operations
+#include <vector>    ///  For std::vector
+
+/**
+ * @namespace operations_on_datastructures
+ * @brief Operations on data structures
+ */
+namespace operations_on_datastructures {
+
+/**
+ * @namespace inorder_successor_of_bst
+ * @brief Functions for the [Inorder successor of a binary search
+ * tree](https://www.youtube.com/watch?v=5cPbNCrdotA) implementation
+ */
+namespace inorder_traversal_of_bst {
+
+/**
+ * @brief A Node structure representing a single node in BST
+ */
+class Node {
+ public:
+    int64_t data;  ///< The key/value of the node
+    Node *left;    ///< Pointer to Left child
+    Node *right;   ///< Pointer to right child
+};
+
+/**
+ * @brief Allocates a new node in heap for given data and returns it's pointer.
+ * @param data Data for the node.
+ * @returns A pointer to the newly allocated Node.
+ * */
+Node *makeNode(int64_t data) {
+    Node *node = new Node();
+    node->data = data;      ///< setting data for node
+    node->left = nullptr;   ///< setting left child as null
+    node->right = nullptr;  ///< setting right child as null
+    return node;
+}
+
+/**
+ * @brief Inserts the given data in BST while maintaining the properties of BST.
+ * @param root Pointer to the root node of the BST
+ * @param data Data to be inserted.
+ * @returns Node* Pointer to the root node.
+ * */
+Node *Insert(Node *root, int64_t data) {
+    if (root == nullptr) {
+        root = makeNode(data);
+    } else if (data <= root->data) {
+        root->left = Insert(root->left, data);
+    } else {
+        root->right = Insert(root->right, data);
+    }
+    return root;
+}
+
+/**
+ * @brief Searches the given data in BST and returns the pointer to the node
+ * containing that data.
+ * @param root Pointer to the root node of the BST
+ * @param data Data to be Searched.
+ * @returns Node* pointer to the found node
+ * */
+Node *getNode(Node *root, int64_t data) {
+    if (root == nullptr) {
+        return nullptr;
+    } else if (root->data == data) {
+        return root;
```

Comment:
Could you please add some documentation like in lines 104 and 107?

Reply:
Yep 👍🏻